internal/pkg/upload/upload.go

```diff
@@ -0,0 +1,286 @@
+	}
+	chunkCount := info.Total / info.ChunkSize
+	if info.Total%info.ChunkSize > 0 {
+		chunkCount += 1
+	}
+	info.Count = int(chunkCount)
+
+	/*
+		Enrich document with additional server-side fields
+	*/
+
+	if err := data.Put(info.ChunkSize, "file", "ChunkSize"); err != nil {
+		return Info{}, err
+	}
+	if err := data.Put(string(StatusAwaiting), "file", "Status"); err != nil {
+		return Info{}, err
+	}
+	if err := data.Put(id, "upload_id"); err != nil {
+		return Info{}, err
+	}
+	if err := data.Put(time.Now().UnixMilli(), "upload_start"); err != nil {
+		return Info{}, err
+	}
+
+	/*
+		Write to storage
+	*/
+	doc, err := json.Marshal(data)
+	if err != nil {
+		return Info{}, err
+	}
+	_, err = CreateFileDoc(ctx, u.bulker, doc, source, docID)
+	if err != nil {
+		return Info{}, err
+	}
+
+	return info, nil
+}
+
+func (u *Uploader) Chunk(ctx context.Context, uplID string, chunkNum int, chunkHash string) (Info, ChunkInfo, error) {
+	// find the upload, details, and status associated with the file upload
+	info, err := u.GetUploadInfo(ctx, uplID)
+	if err != nil {
+		return Info{}, ChunkInfo{}, err
+	}
+
+	/*
+		Verify Chunk upload can proceed
+	*/
+
+	if info.Expired(u.timeLimit) {
+		return Info{}, ChunkInfo{}, ErrUploadExpired
+	}
+	if !info.StatusCanUpload() {
+		return Info{}, ChunkInfo{}, ErrUploadStopped
+	}
+	if chunkNum < 0 || chunkNum >= info.Count {
+		return Info{}, ChunkInfo{}, ErrInvalidChunkNum
+	}
+
+	return info, ChunkInfo{
+		Pos: chunkNum,
+		BID: info.DocID,
+		//FirstReceived: false, // @todo
+		Last: chunkNum == info.Count-1,
+		Size: int(info.ChunkSize),
+		SHA2: chunkHash,
+	}, nil
+}
+
+func validateUploadPayload(info JSDict) error {
+
+	required := [][]string{
+		{"file", "name"},
+		{"file", "mime_type"},
+		{"action_id"},
+		{"agent_id"},
+		{"src"},
+	}
+
+	for _, fields := range required {
+		if value, ok := info.Str(fields...); !ok || strings.TrimSpace(value) == "" {
+			return fmt.Errorf("%s is required", strings.Join(fields, "."))
+		}
+	}
+
+	//@todo: valid action?
+	//@todo: valid src? will that make future expansion harder and require FS updates? maybe just validate the index exists
+
+	if size, ok := info.Int64("file", "size"); !ok {
+		return errors.New("file.size is required")
+	} else if size <= 0 {
+		return fmt.Errorf("invalid file.size: %d", size)
+	}
+	return nil
+}
+
+// Searches for Upload Metadata document in local memory cache if available
+// otherwise, fetches from elasticsearch and caches for next use
+func (u *Uploader) GetUploadInfo(ctx context.Context, uploadID string) (Info, error) {
+	// Fetch metadata doc, if not cached
+	u.mu.RLock()
+	info, exist := u.metaCache[uploadID]
+	u.mu.RUnlock() // not deferred since this must be clear before we gain a write lock below
+	if exist {
+		return info, nil
+	}
+
+	// not found in cache, try fetching
+	info, err := FetchUploadInfo(ctx, u.bulker, uploadID)
+	if err != nil {
+		return Info{}, fmt.Errorf("unable to retrieve upload info: %w", err)
+	}
```

Comment:
just to double check: this would allows for multiple fetches for the same upload id, with misbehaving client, correct?